astropy/nddata/nddata.py

```diff
@@ -5,170 +5,284 @@
-                # one.
-                self._data = np.array(data, subok=True, copy=False)
-                # Quick check to see if what we got out looks like an array
-                # rather than an object (since numpy will convert a
-                # non-numerical input to an array of objects).
-                if self._data.dtype == 'O':
-                    raise TypeError("Could not convert data to numpy array.")
-                self._mask = mask
-            else:
-                self._data = data  # np.array(data, subok=True, copy=False)
-                self._mask = mask
-
-            self._wcs = wcs
-
-            if meta is None:
-                self._meta = OrderedDict()
-            elif not isinstance(meta, collections.Mapping):
-                raise TypeError("meta attribute must be dict-like")
-            else:
-                self._meta = meta
+                    mask = data.mask
+
+                # Just save the data for further processing, we could be given
+                # a masked Quantity or something else entirely. Better to check
+                # it first.
+                data = data.data
 
             if isinstance(data, Quantity):
-                if unit is not None:
-                    raise ValueError("Cannot use the unit argument when data "
-                                     "is a Quantity")
-                else:
-                    self._unit = data.unit
-            else:
-                if unit is not None:
-                    self._unit = Unit(unit)
+                if unit is not None and unit != data.unit:
+                    log.info("Overwriting Quantity's current "
+                             "unit with specified unit")
                 else:
-                    self._unit = None
-            # This must come after self's unit has been set so that the unit
-            # of the uncertainty, if any, can be converted to the unit of the
-            # unit of self.
-            self.uncertainty = uncertainty
+                    unit = data.unit
+                data = data.value
+
+        # Quick check on the parameters if they match the requirements.
+        if (not hasattr(data, 'shape') or not hasattr(data, '__getitem__') or
+                not hasattr(data, '__array__')):
+            # Data doesn't look like a numpy array, try converting it to
+            # one.
+            data = np.array(data, subok=True, copy=False)
+
+        # Another quick check to see if what we got looks like an array
+        # rather than an object (since numpy will convert a
+        # non-numerical/non-string inputs to an array of objects).
+        if data.dtype == 'O':
+            raise TypeError("Could not convert data to numpy array.")
+
+        # Check if meta is a dict and create an empty one if no meta was given
+        if meta is None:
+            meta = OrderedDict()
+        elif not isinstance(meta, collections.Mapping):
```

Comment:
nifty - hadn't realized `collections.Mapping` could be used to check for a dict-like interface.